### ISSUE TITLE
Move scraping of site title and site description to the client (bookmarklet and browser extension)

### ISSUE THREAD

**Issue by eyriewow** [sissbruecker/linkding#118]
Currently, when adding a bookmark through the firefox browser addon the addon will populate certain fields automatically.
I could be wrong here but it seems like the actual scraping of that data, when the bookmark gets added to linkding, is handled by the server. 

Most of the time this is not an issue, but with certain sites like those protected by cloudflare, this leads to unexpected behavior as illustrated here;

Adding a bookmark for a Path of Exile forum post via the browser extension:
Notice the pre-populated title field, as expected
![YoloMouse_PDT1GLgAb7](https://user-images.githubusercontent.com/20689060/116495652-9c478300-a8a3-11eb-9814-fa36ed724886.png)

How that bookmark then appears in linkding:
![YoloMouse_3RlxhE769i](https://user-images.githubusercontent.com/20689060/116495709-bed99c00-a8a3-11eb-917c-2ae5939d9e55.png)

**Comment by sissbruecker** [comment]
For now I would say that this is by design. The scraping happens on the server because:
- it can be reused by the internal bookmark form, by the extension, as well as other tools using the REST API
- fetching a website using AJAX methods from the browser would likely lead to cross origin issues. While the extension might be able to circumvent CORS checks, the internal bookmark form would definitely not

While it's unfortunate that some sites block request coming from servers, I would prefer to keep things simple and keep the logic in one place rather than implement this logic multiple times in different places / languages.

An alternative I can think of is to extend the extension to:
- provide a setting to always set an explicit title + description and get these from the current tab
- ATM the extension only reads the tab title, so it would also need to be extended determine a description from the document

**Comment by sissbruecker** [comment]
Changed the title to include the bookmarklet into this issue. See https://github.com/sissbruecker/linkding/issues/292 for the original request. As mentioned there, if the website metadata is provided by a client, then scraping on the server could be skipped.

I'm more open to this now, as there are bug reports around this from time to time. Ideally the client should provide both the website title and description. Getting the title is straightforward, however the description is not. There are websites (GitHub, Reddit) that do not update the website's meta description tag while navigating through the page, which means the description provided by the client might not be correct. Kind of hard to say which method (client or server scraping) would provide better results on average.

For now I assume server-side scraping is still be better alternative, if someone has ideas around the description issue, feel free to share.

**Comment by joshdick** [comment]
Regarding the description issue, I would love if any currently-selected text on a page would be used as the description when invoking the bookmarklet/extension (the current behavior would be kept if no text is selected.)

Barring that, it would be nice to at least have the ability to manually provide a `description` parameter to `/new` in order to homebrew the functionality described above by customizing the bookmarklet on my own, using it in Apple Shortcuts, etc.

**Comment by sissbruecker** [comment]
The browser extension now allows to use the title and description of the current browser tab instead of fetching those through the server. The bookmarklet still needs to be updated.

**Comment by ccxuy** [comment]
Thanks for your great work!

**Comment by pmorch** [comment]
I've noticed that if I add reddit pages, like e.g. [this r/cats page](https://www.reddit.com/r/cats/comments/1kx7zf9/i_just_took_the_best_photo_in_human_history/) to linkding, I get no title or description.

@sissbruecker wrote above:

> The browser extension now allows to use the title and description of the current browser tab instead of fetching those through the server. The bookmarklet still needs to be updated.

And I'm using the newest [linkding extension - Chrome Web Store](https://chromewebstore.google.com/detail/linkding-extension/beakmhbijpdhipnjhnclmhgjlddhidpe). Shouldn't that pretty much guarantee that there should always be at least a title? Or is that a completely separate issue?

**Comment by ccxuy** [comment]
> I've noticed that if I add reddit pages, like e.g. [this r/cats page](https://www.reddit.com/r/cats/comments/1kx7zf9/i_just_took_the_best_photo_in_human_history/)* to linkding, I get no title or description.

I tested your page, it looks ok.  The title is "I just took the best photo in human history : r/cats"

**Comment by pmorch** [comment]
> I tested your page, it looks ok. The title is "I just took the best photo in human history : r/cats"

That is really weird. It doesn't work for me.

I've even tried with brand new browser instance with:

```
$ google-chrome-stable --user-data-dir=~/tmp/chrome-test
```

And then installed the linkding extension and configured it to talk to my linkding instance. If I click on the linkding icon while on [this r/cats page](https://www.reddit.com/r/cats/comments/1kx7zf9/i_just_took_the_best_photo_in_human_history/) I see this:

<img width="1201" height="592" alt="Image" src="https://github.com/user-attachments/assets/b8ffc3af-0955-415d-86a0-f39763ad9a99" />

No log entries for the linkding docker image or the console for the page.

In order to investigate this further, I've tried to open developer tools for this extension, but I was not able to see the any network request where the extension requests the server for information about this page. In only see:

Some requests under chrome-extension://beakmhbijpdhipnjhnclmhgjlddhidpe/ for index.html and various .js and .css files.

* https://linkding.lightknight.org/api/tags/?limit=5000
* https://linkding.lightknight.org/api/user/profile/

Neither of which contain any information about the current page as far as I can tell.

**Comment by ccxuy** [comment]
“Pre-load page information while browsing” option is on ? @pmorch

**Comment by pmorch** [comment]
> “Pre-load page information while browsing” option is on ?

No, it was not on. When I enabled it I could see the requests to to the /check URL in the extension's background developer tools as soon as the page loads. Thanks for that.

And I can see that the server is indeed consulted and returns empty results for the reddit URL:

```shell
$ curl 'https://linkding.lightknight.org/api/bookmarks/check/?url=https%3A%2F%2Fwww.reddit.com%2Fr%2Fcats%2Fcomments%2F1kx7zf9%2Fi_just_took_the_best_photo_in_human_history%2F' \
  -H 'authorization: Token mytoken' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   198  100   198    0     0    682      0 --:--:-- --:--:-- --:--:--   680
{
  "bookmark": null,
  "metadata": {
    "url": "https://www.reddit.com/r/cats/comments/1kx7zf9/i_just_took_the_best_photo_in_human_history/",
    "title": null,
    "description": null,
    "preview_image": null
  },
  "auto_tags": []
}
```

But works fine for other URLs:
```shell
$ curl 'https://linkding.lightknight.org/api/bookmarks/check/?url=https%3A%2F%2Fwww.morch.com%2Fposts%2F2025-03-22-shpool%2F' \
  -H 'authorization: Token mytoken' | jq  
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   263  100   263    0     0   1267      0 --:--:-- --:--:-- --:--:--  1270
{
  "bookmark": null,
  "metadata": {
    "url": "https://www.morch.com/posts/2025-03-22-shpool/",
    "title": "shpool - much preferred over tmux and screen | Peter V. Mørch's Blog",
    "description": "tmux - but where the terminal works as normal",
    "preview_image": null
  },
  "auto_tags": []
}
```

The docker logs for linkding when the reddit URL is checked is not suspicious:
```
linkding  | [pid: 19|app: 0|req: 10325/24894] ::ffff:172.19.0.1 () {44 vars in 887 bytes} [Fri Aug  8 08:53:33 2025] GET /api/bookmarks/check/?url=https%3A%2F%2Fwww.reddit.com%2Fr%2Fcats%2Fcomments%2F1kx7zf9%2Fi_just_took_the_best_photo_in_human_history%2F => generated 198 bytes in 13 msecs (HTTP/1.1 200) 9 headers in 287 bytes (1 switches on core 0)
```

Am I hijacking a separate issue? Should I open a new one?

**Comment by sissbruecker** [comment]
> I've noticed that if I add reddit pages, like e.g. [this r/cats page](https://www.reddit.com/r/cats/comments/1kx7zf9/i_just_took_the_best_photo_in_human_history/) to linkding, I get no title or description.
> 
> [@sissbruecker](https://github.com/sissbruecker) wrote above:
> 
> > The browser extension now allows to use the title and description of the current browser tab instead of fetching those through the server. The bookmarklet still needs to be updated.
> 
> And I'm using the newest [linkding extension - Chrome Web Store](https://chromewebstore.google.com/detail/linkding-extension/beakmhbijpdhipnjhnclmhgjlddhidpe). Shouldn't that pretty much guarantee that there should always be at least a title? Or is that a completely separate issue?

You specifically need to enable the "Use browser metadata" option in the extension to make it use the data from the current tab instead of getting the metadata through the server.
